### PR TITLE
feat(marketing): add script for sending marketing events from csv data

### PR DESCRIPTION
**Pull request by philbooth**
@rfk, @jbuck, this is a script to process the marketing email event CSV format and send it to Amplitude. There's a fair amount of duplication with the proposed sync script for #36, so we should be able to extract the common stuff as they land.

Should be nothing too surprising hopefully, the only small gotcha was the CSV we received was UTF-16 encoded, so I've added an encoding detection/transform step before piping data to the CSV parser. Happily, unlike the parquet parser, this one operates on a stream so we don't have to download the entire file before processing it.

I'm sticking a `WIP` label on this until we actually have the automated exports dumping to S3 and can verify that part of the code (the S3 bit is untested so far).

Co-authored with @jbuck.

### CODE REVIEW

File: marketing.js [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+const AWS_SECRET_KEY = process.env.FXA_AWS_SECRET_KEY
+const AWS_S3_BUCKET = '*** TODO ***'
+const AWS_S3_PREFIX = '*** TODO ***'
+
+const MAX_EVENTS_PER_BATCH = 10
+const HMAC_KEY = process.env.FXA_AMPLITUDE_HMAC_KEY
+const API_KEY = process.env.FXA_AMPLITUDE_API_KEY
+
+if (process.argv.length !== 3) {
+  console.error(`Usage: ${process.argv[1]} {YYYY-MM-DD-hh-mm | LOCAL PATH}`)
```
Comment (philbooth) [inline marketing.js]: Not sure what frequency the data will be exported to S3 with yet, but I've optimistically written this to cope with per-minute cadence.

File: marketing.js [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+  return processStream(fs.createReadStream(process.argv[2]))
+}
+
+function processStream (stream) {
+  let eventCount = 0, batch = [], error
+
+  return new Promise((resolve, reject) => {
+    stream
+      .pipe(new AutoDetectDecoderStream())
```
Comment (philbooth) [inline marketing.js]: The example CSV we were sent was UTF-16. This step detects the encoding using the BOM if present (it was in the sample data) and transforms to JS strings.

File: marketing.js [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+fs.unlinkAsync = Promise.promisify(fs.unlink)
+
+const TIMESTAMP = /^(20[1-9][0-9])-([01][0-9])-([0-3][1-9])-([0-2][0-9])-([0-5][0-9])$/
+const EVENT = /^mktg-([a-z]+_[a-z]+)$/
+
+const AWS_ACCESS_KEY = process.env.FXA_AWS_ACCESS_KEY
+const AWS_SECRET_KEY = process.env.FXA_AWS_SECRET_KEY
+const AWS_S3_BUCKET = '*** TODO ***'
+const AWS_S3_PREFIX = '*** TODO ***'
```
Comment (philbooth) [inline marketing.js]: ...and these are the other reason it's still `WIP`, obviously.

File: marketing.js [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+  if (! eventType) {
+    return
+  }
+
+  const time = getTime(row)
+  if (! time || time < 0) {
+    return
+  }
+
+  const user_id = row.FXA_ID
```
Comment (philbooth) [inline marketing.js]: Whoops, just realised I'm not hashing this. Will fix.

**Comment by davismtl** [comment]
@philbooth ,  @bniolet1 is wondering if you and @jbuck could tell him where / how he should send his data dumps to S3.

**Comment by philbooth** [comment]
> @bniolet1 is wondering if you and @jbuck could tell him where / how he should send his data dumps to S3.

Not sure, I think @jrgm is on the case. There were some notes from the back-end meeting on Tuesday:

> created bucket, pinged bniolet on irc, need to check if he controls user fxa-exacttarget

**Comment by philbooth** [comment]
[Sounds like](https://bugzilla.mozilla.org/show_bug.cgi?id=1409247#c45) this won't be happening until late April at the earliest, so I'm moving this back to `next`.

**Comment by vladikoff** [comment]
from mtg: we have some sample data, not flow data. We can verify the S3 part

**Comment by philbooth** [comment]
@jbuck, thanks for the fixes! I pushed a couple more tweaks on top of your changes, one to make lint pass and one to prefer underscores to hyphens in the resulting event name (underscores are what's specced in the taxonomy spreadsheet).

**Comment by philbooth** [comment]
@jbuck, I noticed that the stage Amplitude project contains marketing sent events but not open or click events. Is that because we just tried a small sample that didn't have the other event types or is it because we need to tweak the event matching regex to pick them up?

If it's the former, we should run a bigger test just to make sure that we're mapping all of the required event types correctly. If it's the latter, let's fix it (obviously).

**Comment by bniolet1** [comment]
@philbooth I'm guessing this is due to the fact that there are two file types sent over to Brick. One is all "blast" sends and the other file contains all other activities (opens, clicks and triggered sends). The "Sent" file is much smaller and only appears sporadically due to the fact that we don't send a blast email every day. 

For more details, see [comment 52](https://bugzilla.mozilla.org/show_bug.cgi?id=1409247#c52) in the bug.

**Comment by jbuck** [comment]
@philbooth I think we should merge this PR as-is, since it's working quite well for files from the local filesystem.

I need to run some tests to see if Lambda is fast enough to submit data to Amplitude, otherwise I'll run this script as a cron job on the fxa-admin host.

**Pull request merged**